---
output: github_document
---

```{r echo = FALSE, message = FALSE}
my_logical <- c(T,T,T,F,F)
library(tidyverse)
setwd("C:/Dropbox/GitHub/Practical_computing/Class_10")
```

# Class 10
# `dplyr`

## `dplyr` and `tidyverse`

`dplyr` is a package from `tidyverse` ecosystem (a set of packages sharing a common philosophy), which has been enjoying an enormous popularity in recent years. Sometimes even one has an impression that people equate R with `tidyverse`. The main purpose of `dplyr` is to simplify work with data frames.  In this course we first introduced base R, because we are convinced that its understanding is critical for efficient work in R. During this class you'll learn how to use `dplyr` and we'll go through a real-life example of filtering and summarizing a complex, "real", data.
For more info on `tidyverse` see [webpage](https://www.tidyverse.org/) and [free book](https://r4ds.had.co.nz/)

## Six `dplyr` verbs

We'll illustrate the basic `dplyr` functions (verbs) using the built-in `iris` dataset that contains measurements (in cm) of several flower traits of several *Iris* species.

#### Exercise 1
Display structure and summary of `iris` dataset using `str()` and `summary()`. Answer the following questions:

* which traits were measured?
* how many flowers were measured in total?
* which *Iris* species were studied?
* what were sample sizes of particular species?

Here's a schematic view of a flower with the most important structures labeled: 

![flower](flower.png)

---

All six functions work similarly:

1. The first argument is always a data frame.
2. Subsequent arguments describe what the function is supposed to do with the data frame. You can refer to the columns of the data frame without using quotation marks.
3. The result is a data frame.

#### Exercise 2
Load `dplyr` package

---

### Filtering rows with `filter()`
`filter()` returns rows that fulfill a logical condition. You can combine any number of conditions.

#### `filter()` examples

Display data for *Iris setosa*
```{r, eval = FALSE}
filter(iris, Species == "setosa")
```

```{r, echo = FALSE}
filter(iris, Species == "setosa") %>% head(n = 10)
```


**Note:**

 * we used double equal sign (`==`) to specify a comparison
 * the column name was not enclosed in quotation marks, while the value (species name, character string) was.

Display all observations for *Iris setosa* that have sepal length at least 5.5 cm:

```{r, eval = FALSE}
filter(iris, Species == "setosa", Sepal.Length >= 5.5)
```

```{r, echo = FALSE}
filter(iris, Species == "setosa", Sepal.Length >= 5.5) %>% head(n = 10)
```

If both conditions are to be met, they can be separated by comma (`,`) or by the AND sign (`&`):

```{r}
filter(iris, Species == "setosa" & Sepal.Length >= 5.5)
```

You can use alternatives and group conditions using parentheses. To display all observations of *I. versicolor* and *I. virginica* that have petal length at least 5 cm:

```{r, eval = FALSE}
filter(iris, (Species == "versicolor" | Species == "virginica"), Petal.Length >=5)
```

or

```{r,, eval = FALSE}
filter(iris, (Species == "versicolor" | Species == "virginica") & Petal.Length >=5)
```

```{r, echo = FALSE}
filter(iris, (Species == "versicolor" | Species == "virginica") & Petal.Length >=5) %>% 
  head(n = 10)
```

Note, that the following doesn't work:

```{r, error = TRUE}
filter(iris, (Species == "versicolor" | "virginica"), Petal.Length >=5)
```


#### Exercise 3
Select all observations, regardless of species, that have sepal length smaller than 6 cm or petal length smaller than 5 cm, and sepal width larger than 4 cm. 

Expected result:  
```{r, echo = FALSE}
filter(iris, (Sepal.Length < 6 | Petal.Length > 5) & Sepal.Width > 4)
```

---

### Ordering rows with `arrange()`
`arrange()`returns data frame sorted according to a column or a combination of columns. If you provide more than one column, each additional column will be used to break the ties. If you want to sort according to a column in descending order, use `desc()`. Note, that numbers stored as character are sorted differently than numbers stored as numeric, and when sorting logical values `FALSE` (`0`) comes before `TRUE` (`1`).


#### `arrange()` example

Sort the `iris` dataset according to species in alphabetical (ascending) order and within the species from the longest to the shortest sepal

```{r, eval = FALSE}
arrange(iris, Species, desc(Sepal.Length))
```

#### Exercise 4
Sort species in descending order, and observations within each species according to the increasing petal length.

Expected result:
```{r, echo = FALSE}
head(arrange(iris, desc(Species), Petal.Length), n = 10)
```

---

### Picking, dropping and re-ordering columns with `select()`
Select selects, drops or re-arranges columns of a data frame. Columns can be referred to by name, position in the data frame or by an expression.

> #### `select` helpers
The following helper functions can be used to select columns that match a pattern:
>
`starts_with()` selects columns that start with a string  
`ends_with()` selects columns that end with a string  
`contains()` selects columns that contain a string  
`matches()` selects columns matching a regular expression  
>
There are also useful helpers that allow to select column on the basis of a character vector containing names. See `?select` for more details.   

#### `select()` examples

Select `iris` columns 1, 2, 3 and 5:
```{r, eval = FALSE}
select(iris, 1:3, 5)
```

or, just drop column 4:

```{r, eval = FALSE}
select(iris, -4)
```

```{r, echo = FALSE}
select(iris, -4) %>% head(n = 10)
```

Select species name and petal measurements:
```{r, eval = FALSE}
select(iris, Species, starts_with("Petal"))
```

```{r, echo = FALSE}
select(iris, Species, starts_with("Petal")) %>% head(n = 10)
```

The example above shows that `select()` can be used to reorder variables. A useful trick to move a single column (here `Species`) to the beginning of the data frame:
```{r, eval = FALSE}
select(iris, Species, everything())
```

```{r, echo = FALSE}
select(iris, Species, everything()) %>% head(n = 10)
```

#### Exercise 5
Select width measurements and species from `iris`, at the same time relocating species to the beginning of the data frame.  
Expected result:

```{r, echo = FALSE}
head(select(iris, Species, ends_with("Width")), n = 10)
```

---

#### Exercise 6
Drop petal measurements from `iris` data frame.  
Expected result:

```{r, echo = FALSE}
head(select(iris, -(starts_with("Petal"))), n = 10)
```

---

### Creating new variables as functions of the existing ones with `mutate()`
`mutate()` adds a new column at the end of the data frame. The value of this column can be a single element vector provided by the user or, usually more usefully, a formula that uses values of other variables. Once you create the new column, you can immediately use it in the same `mutate()` call.

#### `mutate` examples

Add to `iris`a new variable called `One` with the value `1` and the data type character for all observations.
```{r, eval = FALSE}
mutate(iris, One = "1")
```

```{r, echo = FALSE}
mutate(iris, One = "1") %>% head(n = 10)
```

Create new variable named `Petal.Ratio`, the value of which will be the ratio of petal length to petal width:
```{r, eval = FALSE}
mutate(iris, Petal.Ratio = Petal.Length/Petal.Width)
```

```{r, echo = FALSE}
mutate(iris, Petal.Ratio = Petal.Length/Petal.Width) %>% head(n = 10)
```


#### Exercise 7
Create, using a single `mutate()` call, two new variables: Petal.Length.Squared, Sepal.Length.Squared, containing the squared length of petal and sepal, respectively.  
Expected result:  
```{r, echo = FALSE}
head(mutate(iris, Petal.Length.Squared = Petal.Length^2,
            Sepal.Length.Squared = Sepal.Length^2), n = 10)
```

---

### Grouping cases by variable(s) with `group_by()`

You could (rightly) consider some of the preceding examples and exercises boring, so let's move to something more exciting.  `group_by()` creates a **grouped data frame**, that is a data frame which rows are assigned to various groups based on the value of one or more grouping variables. The data frame itself is not changed, but the way operations on the data frame are performed is. For example, if you use a function, such as `mean()` with `mutate()`, the mean of the group to which a row belongs will be used for this row.

#### `group_by()` examples

Group `iris` by species
```{r, eval = FALSE, warning = FALSE}
group_by(iris, Species)
```

When you execute the command above, you'll see that the object is not anymore `data.frame`, but now it's called `tibble` and you'll find information about grouping added, though no information in the table was altered:
```{r, echo = FALSE, warning = FALSE}
group_by(iris, Species)
```

#### Exercise 8
Group `iris` by species, assign the grouped dataset to variable, and use mutate to add column `Mean.Sepal.Length`, which would contain the mean value of sepal length. Are all values in this column identical? Why?

---

#### The pipe operator (`%>%`)

Usually, when you want to use several functions in sequence, you repeatedly assign the result to a variable, which is fine, but a bit inconvenient:
```{r}
a <- select(iris, Species, starts_with("Sepal"))
a <- mutate(a, Sepal.Ratio = Sepal.Length/Sepal.Width)
head(a)
```

Here's where the **pipe** `%>%` comes handy. Although the symbol used is different, its behaviour is similar to the pipe operator `|` of the Linux shell. In conjunction with `dplyr` verbs it allows creating pipelines without the need of assigning intermediate results to variables. Note, that when a `dplyr` function is used following `%>%`, you don't specify the data frame the function operates on - because the data frame is passed by `%>%`!

So, instead of the code above you can use:
```{r}
a <- select(iris, Species, starts_with("Sepal")) %>% mutate(Sepal.Ratio = Sepal.Length/Sepal.Width)
head(a)
```
or
```{r}
a <- iris %>% select(Species, starts_with("Sepal")) %>% mutate(Sepal.Ratio = Sepal.Length/Sepal.Width)
head(a)
```

Using `%>%` not only eliminates the need for intermediate variables, but also makes code more **readable**. In this case, you first take `iris`, then select some columns and then add a column based on the values of the existing columns. The way your code is written reflects the sequence of steps.

Before we move on, we'll show a useful technique that allows, for example, easy standardization. Our task is to standardize the values of sepal length by subtracting each from the species mean and dividing the result by the species standard deviation (this is called Z-standardization). Here's the code:


```{r, warning = FALSE}
iris %>% select(Species, Sepal.Length) %>%  group_by(Species) %>%
  mutate(sp.Mean = mean(Sepal.Length),
         sp.SD = sd(Sepal.Length),
         Zstand.Sepal.Length = (Sepal.Length - sp.Mean)/sp.SD)
```

we could use `select()` again to drop the intermediate variables we no longer need:
```{r, warning = FALSE}
iris %>% select(Species, Sepal.Length) %>%  group_by(Species) %>%
  mutate(sp.Mean = mean(Sepal.Length),
         sp.SD = sd(Sepal.Length),
         Zstand.Sepal.Length = (Sepal.Length - sp.Mean)/sp.SD) %>% select(-c(sp.Mean, sp.SD))
```

Note, that the columns to drop were passed as a vector of column names not enclosed in quotation marks (quotation marks are allowed, but not necessary).

### Summarising with `summarise()`
When used on a data frame that is not grouped `summarise` just applies a function to a column and returns a small data frame containing the result:

```{r}
summarise(iris, mean.Sepal.Length = mean(Sepal.Length))
```

This may be useful, but rather not terribly so.

Things change when you use summarise with a grouped data frame. Then, writing very little code you can get lots of useful results:
```{r, warning=FALSE}
iris %>% select (Species, Petal.Length) %>% group_by(Species) %>% 
  summarise(min.Petal.Length = min(Petal.Length),
            mean.Petal.Length = mean(Petal.Length), 
            max.Petal.Length = max(Petal.Length))
```


## Practical `dplyr`

In the remaining part of the class we'll illustrate the practical use of `dplyr` with a real life example. Our task will be to summarize data on MHC class I genotypes from many populations of  several *Triturus* newt species. These genes are duplicated (with different number of genes in different individual) and highly polymorphic, so each individual has several alleles, and the total number of alleles in a species go into hundreds.  

We will calculate:

* For each population:
  * number of analysed individuals (`n_ind`)
  * total number of MHC alleles (`n_all`)
  * mean number of MHC alleles per individual (`mean_n_all_ind`)
* For each species you will calculate the same three summaries and additionally also:
  * number of sampled populations (`n_pop`)
  
We'll use the names indicated above as column names of your final result data frame.

* **Download data**  
[Here](https://www.dropbox.com/sh/6s60oyt4snc8fgj/AAAyR9wv3t9hWUah2bq_aNtwa?dl) is the folder with the data we will use. Please download the files to your computer. All files are encoded in UTF-8. Have a look at the content of the files in Notepad++:
    * `genotypes.txt` - tab-delimited file, 1st row contains column names, the remaining rows contain individual IDs and genotypes encoded as presence (`1`) or absence (`0`) of each allele. Allele names are provided in the 1st row.
    * `ID_locality_species.txt`- tab-delimited file containing individual IDs (as in `genotypes.txt` but more), locality name, and species designation. There is also an additional column `transctiptome` that we will not use. Note that locality names contain characters outside the standard Latin character set.
    * `localities.txt`- tab-delimited file containing various information about localities. We'll need only a subset of localities and a subset of columns.
    
    Organisation of these data into the three files emphasizes the concepts we learned during this course:
    * genotype file contains only information about individual IDs and genotypes.
    * further information about each individual is contained in another file, and can be linked with genotypes when needed using the column `ID` which is shared between the files.
    * another file contains information about localities, which can be linked with information about individuals using the shared `locality` column.
    * `ID_locality_species.txt` and `localities.txt` contain more rows and columns than needed. This is not a problem, because we will extract only the information we need from these files.


* **Set up new project in RStudio**  
Within the project folder create new directory named `data`. Copy the three previously downloaded files to `data`

* **Create new R script**  
Make sure it's in you project folder (NOT in `data`). In this script you'll save all the commands we'll use below.

* **Load `tidyverse`**  
```{r, eval = FALSE}
library(tidyverse)
```
  We're loading the entire `tidyverse` collection, because we'll need some functions that are in packages other than `dplyr`

* **Read genotypes**  
```{r}
gen <- read.table("data/genotypes.txt", header = TRUE, sep = "\t", encoding = "UTF-8")
head(gen[,1:10])
```
We used several options to modify the default behaviour or `read.table()`:  

  * column names to be taken from the 1st row,
  * columns are separated by tabs
  * text is encoded as UTF-8

* **Check whether any individual IDs are duplicated**
```{r}
nrow(gen)
gen %>% select(ID) %>% distinct() %>% nrow()
```
Here we counted the number of rows in the original dataframe and the dataframe containing only unique IDs

* **Filter out missing data**  
The way genotypes were obtained ensures that if information for an individual is missing for any allele, it's actually mising for all. So it's enough to filter out rows that contain `NA` for the first allele. 
```{r}
gen_noNA <- gen %>% filter(!is.na(i_0003))
```
Note that we used a dedicated function `is.na()` to identify `NA` values.

* **Read information about each individual's locality and species**  
We'll need only two columns, so we drop `transcriptome` on the way
```{r}
ID_info <- read.table("data/ID_locality_species.txt", header = TRUE, sep = "\t", encoding = "UTF-8") %>% 
  select(-transcriptome)
head(ID_info)
```

* **Read information about localities**  
We'll need just the locality name, country and geographic coordinates, so we select only these four columns:
```{r}
localities <- read.table("data/localities.txt", header = TRUE, sep = "\t", encoding = "UTF-8") %>% 
  select(locality, country, latitude, longitude)
head(localities)
```

* **Reshape the genotype data into the "long" format**  
In this step we will radically reshape our data, which will simplify further tasks. Instead of having each allele as a separate column, we'll collapse these columns into two: i) allele name, ii) information about presence (`1`) or absence (`0`) in a given individual.  
So, we take our data frame that looks like this:
```{r}
head(gen[,1:20], n = 20)
```
And reshape it using the following command:
```{r, warning = FALSE}
gen_long <- gen_noNA %>% pivot_longer(-ID, names_to = "allele", values_to = "present_absent")
head(gen_long, n = 20)
```
We called `pivot_longer()`, informing the function that we want to reshape all the columns except `ID`, that names of the columns should go the new `allele` column, and that values from these columns should go to the `present_absent` column.
Let's compare the number of rows in the original and reshaped data frame:
```{r}
nrow(gen_noNA)
nrow(gen_long)
```
In the long format, the information that an individual doesn't have a particular allele is not really useful, so we can just get rid of the rows with (`0`) in `present_absent`. Then, `present_absent` column would not be informative anymore, so we can drop it:
```{r}
gen_long <- gen_long %>% filter(present_absent == 1) %>% select(-present_absent)
nrow(gen_long)
```
It's much shorter now!

* **Add information about populations and species to the genotype table**  
We start be adding information about locality and species:
```{r, warning = FALSE}
gen_long <- gen_long %>% left_join(ID_info, by = "ID")
head(gen_long, n = 15)
```
Here we used `left_join()` to add columns from `ID_info` data frame to `gen_long` data frame. Left join means that in the resulting data frame all rows from the left (i.e., `gen_long`) data frame will be retained, and only these rows from `ID_info` data frame will be added, which match a value of `ID` column in any row of `gen_long` data frame. If multiple rows from `ID_info` match a single row of `gene_long`, the resulting data frame will have more rows than `gen_long`.

And now we'll add information about localities:
```{r, warning = FALSE}
gen_long <- gen_long %>% left_join(localities, by = "locality")
head(gen_long, n = 15)
```

* **Calculate summaries we want**  
First, we calculate for each population: 
    * sample size
    * the number of distinct alleles
    * mean number of alleles per individual
```{r, warning=FALSE, message=FALSE}
pop_sum <- gen_long %>% group_by(locality, species) %>% summarise(n_ind = n_distinct(ID),
                                                                  n_all = n_distinct(allele),
                                                                  mean_n_all_ind = n()/n_ind)
head(pop_sum)
```
`n_distinct()` calculates the number of different values in a data frame column **within each group**, as defined in `group_by()`  
`n()` counts the number of rows in each group.  
Please stop here to make sure that you understand what and how was calculated here.

In principle, it's posible that more than one species occurs in a locality, let's check it!

```{r}
nrow(pop_sum)
pop_sum %>% distinct(locality) %>% nrow()
```



Then, we calculate the same statistics, but with data grouped by species:
```{r, warning = FALSE, message=FALSE}
spec_sum <- gen_long %>% group_by(species) %>% summarise(n_ind = n_distinct(ID),
                                                         n_all = n_distinct(allele),
                                                         mean_n_all_ind = n()/n_distinct(allele),
                                                         n_pop = n_distinct(locality))
head(spec_sum)
```

We're done now! But, perhaps, you'd like to have a single table with all your summaries, so that the species summary is below the list of populations of each species for the ease of inspection.

* **Combine the two result data frames into one**

```{r, warning = FALSE}
sum_sum <- pop_sum %>% bind_rows(spec_sum) 
sum_sum
```
`bind_rows()` added rows from `spec_sum` to the end of `pop_sum`. Because the data frames differed in the number of columns, i.e., `n_pop` was missing from `pop_sum` and `locality` was missing from `spec_sum`, the columns from both data frames were combined and filled with `NA` where needed.  
Have a look at the entire data frame in RStudio viewer (`View()`).

Now, we'll make use of the fact that `NA` values are put at the end of a column sorted with `arrange()`, to place summary for each species under the summaries of its populations:
```{r, warning = FALSE}
sum_sum <- sum_sum %>% arrange(species, locality)
sum_sum
```

And finally, we'd like to replace `NA` in the `Locality` column with the word "Overall"
```{r, warning = FALSE}
sum_sum <- sum_sum %>% 
  mutate(locality = ifelse(is.na(locality), "Overall", locality))
sum_sum
```

`ifelse()` is a function that operates on each row of the data frame, so its useful together with mutate. If the logical condition is met, i.e., if there is `NA` in the `loaclity` column of a given row, the value of the new column `locality` in this row will be `"Overall"`, otherwise, the value will just be taken from the eisting `locality` column.

* **Save the resulting table into a text file**  
We want tab-delimited text file without enclosing any values in quotes
```{r, eval = FALSE}
write.table(sum_sum, "MHC_summary.txt", quote = FALSE, sep = "\t")
```
The same can be done by using `dplyr` function `write_tsv()` called with default options:
```{r, eval = FALSE}
write_tsv(sum_sum, "MHC_summary.txt")
```

### Puting the code together.  
Let's see now how much code was actually needed to accomplish our task. Below is the code streamlined a little bit by a more extensive use of `%>` and skipping commands that print intermediate results. Copy this code to your script and write comments (comment lines start with `#`) explaining what particular commands do.

```{r, eval = FALSE}
library(tidyverse)

gen <- read.table("data/genotypes.txt", header = TRUE, sep = "\t", encoding = "UTF-8")

nrow(gen)
gen %>% select(ID) %>% distinct() %>% nrow()

gen_noNA <- gen %>% filter(!is.na(i_0003))

ID_info <- read.table("data/ID_locality_species.txt", header = TRUE, sep = "\t", encoding = "UTF-8") %>% select(-transcriptome)

localities <- read.table("data/localities.txt", header = TRUE, sep = "\t", encoding = "UTF-8") %>% 
  select(locality, country, latitude, longitude)

gen_long <- gen_noNA %>% pivot_longer(-ID, names_to = "allele", values_to = "present_absent") %>% 
  filter(present_absent == 1) %>% select(-present_absent)

gen_long <- gen_long %>% left_join(ID_info, by = "ID") %>% 
  left_join(localities, by = "locality")

pop_sum <- gen_long %>% group_by(locality, species) %>% summarise(n_ind = n_distinct(ID),
                                                                  n_all = n_distinct(allele),
                                                                  mean_n_all_ind = n()/n_ind)
                                                
spec_sum <- gen_long %>% group_by(species) %>% summarise(n_ind = n_distinct(ID),
                                                         n_all = n_distinct(allele),
                                                         mean_n_all_ind = n()/n_distinct(allele),
                                                         n_pop = n_distinct(locality))

sum_sum <- pop_sum %>% bind_rows(spec_sum) %>% arrange(species, locality) %>% 
  mutate(locality = ifelse(is.na(locality), "Overall", locality))

write_tsv(sum_sum, "MHC_summary.txt")
```

